### PR TITLE
Find runner groups that visible to repository using a single API call.

### PR DESCRIPTION
**Pull request by TingluoHuang**
The [ListRunnerGroup API](https://docs.github.com/en/rest/reference/actions#list-self-hosted-runner-groups-for-an-organization) now add a new query parameter `visible_to_repository`.

We were doing `N+1` lookup when trying to find which runner group can be used for job from a certain repository.
- List all runner groups
- Loop through all groups to check repository access for each of them via [API](https://docs.github.com/en/rest/reference/actions#list-repository-access-to-a-self-hosted-runner-group-in-an-organization)

The new query parameter `visible_to_repository` should allow us to get the runner groups with access in one call.

Limitation:
- The new query parameter is only supported in GitHub.com, which means anyone who uses ARC in GitHub Enterprise Server won't get this.
- I am working on a PR to update `go-github` library to support the new parameter, but it will take a few weeks for a newer `go-github` to be released, so in the meantime, I am duplicating the implementation in ARC as well to support the new query parameter.

### CODE REVIEW

File: simulator/runnergroup_visibility.go [hunk 1]
```diff
@@ -18,30 +18,47 @@ func (c *Simulator) GetRunnerGroupsVisibleToRepository(ctx context.Context, org,
 		panic(fmt.Sprintf("BUG: owner should not be empty in this context. repo=%v", repo))
 	}
 
-	runnerGroups, err := c.Client.ListOrganizationRunnerGroups(ctx, org)
-	if err != nil {
-		return visible, err
-	}
+	if c.Client.GithubBaseURL == "https://github.com/" {
```
Comment (TingluoHuang) [inline simulator/runnergroup_visibility.go:21]: Is there any better way to detect whether we configure with GitHub Enterprise server?

Comment (TingluoHuang) [reply simulator/runnergroup_visibility.go:21]: Can I feature flag these changes in case they break any customers?

File: github/github.go [hunk 2]
```diff
@@ -286,6 +309,30 @@ func (c *Client) ListRunnerGroupRepositoryAccesses(ctx context.Context, org stri
 	return repos, nil
 }
 
+// listOrganizationRunnerGroupsVisibleToRepo lists all self-hosted runner groups configured in an organization which can be used by the repository.
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/actions#list-self-hosted-runner-groups-for-an-organization
+func (c *Client) listOrganizationRunnerGroupsVisibleToRepo(ctx context.Context, org, repo string, opts *github.ListOptions) (*github.RunnerGroups, *github.Response, error) {
+	u := fmt.Sprintf("orgs/%v/actions/runner-groups?visible_to_repository=%v&per_page=%v", org, repo, opts.PerPage)
```
Comment (TingluoHuang) [inline github/github.go:316]: I think I need to handle the case where `opts` is nil...

**Comment by TingluoHuang** [comment]
I tested the change out for the runner group that the repository has access to.
```
2022-04-20T02:23:43Z    DEBUG   controllers.webhookbasedautoscaler      Searching in runner groups      {"event": "workflow_job", "hookID": "354156966", "delivery": "e5aefd50-c050-11ec-961b-dd81adafddd0", "workflowJob.status": "completed", "workflowJob.labels": ["self-hosted", "arc-runner", "tingluohuang-arc-setup-g7v5994fv4jx"], "repository.name": "public-repo", "repository.owner.login": "enterprise-canary", "repository.owner.type": "Organization", "enterprise.slug": "test-enterprise", "action": "completed", "groups": "RunnerGroup{Scope:Organization, Kind:Custom, Name:test_group}"}
2022-04-20T02:23:43Z    DEBUG   controllers.webhookbasedautoscaler      groups  {"event": "workflow_job", "hookID": "354156966", "delivery": "e5aefd50-c050-11ec-961b-dd81adafddd0", "workflowJob.status": "completed", "workflowJob.labels": ["self-hosted", "arc-runner", "tingluohuang-arc-setup-g7v5994fv4jx"], "repository.name": "public-repo", "repository.owner.login": "enterprise-canary", "repository.owner.type": "Organization", "enterprise.slug": "test-enterprise", "action": "completed", "groups": "RunnerGroup{Scope:Organization, Kind:Custom, Name:test_group}"}
2022-04-20T02:23:43Z    DEBUG   controllers.webhookbasedautoscaler      Found 1 HRAs by key     {"key": "enterprise-canary/group/test_group"}
2022-04-20T02:23:43Z    DEBUG   controllers.webhookbasedautoscaler      job scale up target found       {"event": "workflow_job", "hookID": "354156966", "delivery": "e5aefd50-c050-11ec-961b-dd81adafddd0", "workflowJob.status": "completed", "workflowJob.labels": ["self-hosted", "arc-runner", "tingluohuang-arc-setup-g7v5994fv4jx"], "repository.name": "public-repo", "repository.owner.login": "enterprise-canary", "repository.owner.type": "Organization", "enterprise.slug": "test-enterprise", "action": "completed", "enterprise": "test-enterprise", "organization": "enterprise-canary", "repository": "public-repo", "key": "enterprise-canary/group/test_group"}
```

Later I updated the runner group setting to unselect my test repository.
```
2022-04-20T02:26:05Z    DEBUG   controllers.webhookbasedautoscaler      Searching in runner groups      {"event": "workflow_job", "hookID": "354156966", "delivery": "39f6ead0-c051-11ec-8d97-71f09df3dd76", "workflowJob.status": "queued", "workflowJob.labels": ["self-hosted", "arc-runner", "tingluohuang-arc-setup-g7v5994fv4jx"], "repository.name": "public-repo", "repository.owner.login": "enterprise-canary", "repository.owner.type": "Organization", "enterprise.slug": "test-enterprise", "action": "queued", "groups": ""}
2022-04-20T02:26:05Z    DEBUG   controllers.webhookbasedautoscaler      groups  {"event": "workflow_job", "hookID": "354156966", "delivery": "39f6ead0-c051-11ec-8d97-71f09df3dd76", "workflowJob.status": "queued", "workflowJob.labels": ["self-hosted", "arc-runner", "tingluohuang-arc-setup-g7v5994fv4jx"], "repository.name": "public-repo", "repository.owner.login": "enterprise-canary", "repository.owner.type": "Organization", "enterprise.slug": "test-enterprise", "action": "queued", "groups": ""}
2022-04-20T02:26:05Z    DEBUG   controllers.webhookbasedautoscaler      no repository/organizational/enterprise runner found    {"event": "workflow_job", "hookID": "354156966", "delivery": "39f6ead0-c051-11ec-8d97-71f09df3dd76", "workflowJob.status": "queued", "workflowJob.labels": ["self-hosted", "arc-runner", "tingluohuang-arc-setup-g7v5994fv4jx"], "repository.name": "public-repo", "repository.owner.login": "enterprise-canary", "repository.owner.type": "Organization", "enterprise.slug": "test-enterprise", "action": "queued", "repository": "enterprise-canary/public-repo", "organization": "enterprise-canary", "enterprise": "test-enterprise"}
2022-04-20T02:26:05Z    DEBUG   controllers.webhookbasedautoscaler      Scale target not found. If this is unexpected, ensure that there is exactly one repository-wide or organizational runner deployment that matches this webhook event {"event": "workflow_job", "hookID": "354156966", "delivery": "39f6ead0-c051-11ec-8d97-71f09df3dd76", "workflowJob.status": "queued", "workflowJob.labels": ["self-hosted", "arc-runner", "tingluohuang-arc-setup-g7v5994fv4jx"], "repository.name": "public-repo", "repository.owner.login": "enterprise-canary", "repository.owner.type": "Organization", "enterprise.slug": "test-enterprise", "action": "queued"}
```

**Review by mumoshu — APPROVED** [review]
LGTM. It's definitely looking much better and this degree of improvement can't be done with an enhancement in the API layer! I greatly appreciate your collaboration on this. Thank you!

**Pull request merged**